admission: split plugins into validating-only and mutating+validating

sttts opened the pull request:
Prerequisite to enable admission webhooks by default.

liggitt commented on pkg/project/apiserver/admission/nodeenv/admission.go:
```diff
@@ -67,12 +69,22 @@ func (p *podNodeEnvironment) Admit(a admission.Attributes) (err error) {
+	if mutationAllowed {
+		// modify pod node selector = project node selector + current pod node selector
+		pod.Spec.NodeSelector = labelselector.Merge(projectNodeSelector, pod.Spec.NodeSelector)
+	}
```
needs an else to fail if the merged selector is not identical to the pod selector, right?

sttts replied:
hah, you were 1 min faster :)

sttts replied:
done

sttts commented on pkg/project/apiserver/admission/nodeenv/admission.go:
```diff
@@ -67,12 +69,22 @@ func (p *podNodeEnvironment) Admit(a admission.Attributes) (err error) {
-	pod.Spec.NodeSelector = labelselector.Merge(projectNodeSelector, pod.Spec.NodeSelector)
+	if mutationAllowed {
+		// modify pod node selector = project node selector + current pod node selector
+		pod.Spec.NodeSelector = labelselector.Merge(projectNodeSelector, pod.Spec.NodeSelector)
```
should we check that the project node selector is contained?

sttts replied:
on validation I mean.

sttts replied:
yes, old question.

liggitt commented on pkg/build/apiserver/admission/jenkinsbootstrapper/admission.go:
```diff
@@ -58,7 +59,7 @@ func NewJenkinsBootstrapper() admission.Interface {
 }
 
-func (a *jenkinsBootstrapper) Admit(attributes admission.Attributes) error {
+func (a *jenkinsBootstrapper) Validate(attributes admission.Attributes) error {
```
This one will be fun for dry run

liggitt commented on pkg/security/apiserver/admission/sccadmission/scc_exec.go:
```diff
@@ -25,21 +25,21 @@ func RegisterSCCExecRestrictions(plugins *admission.Plugins) {
 	client              kclientset.Interface
 }
 
-func (d *sccExecRestrictions) Admit(a admission.Attributes) (err error) {
```
Double check validating admission gets called on exec requests… it's not a normal storage impl

liggitt replied:
Do we have an e2e test that checks if this is active and effective?

sttts replied:
`Validate` is called: https://github.com/kubernetes/kubernetes/blob/d1636b8019fa042eb1135263b00293b1c806b1d7/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go#L132

I added an integration test for SCCExecRestrictions.

liggitt commented:
/lgtm
/hold
for question on sccexec

liggitt commented:
also looks like relevant test failures

sttts commented:
Have extended the PR quite a lot by adding `Validate` funcs where we only had `Admit`. Also added tests for both `Admit` and `Validate` where it made sense.

sttts commented:
@simo5 @enj can you take a look whether the changes are sane?

simo5 commented on pkg/build/apiserver/admission/secretinjector/admission.go:
```diff
@@ -93,7 +104,13 @@ func (si *secretInjector) Admit(attr admission.Attributes) (err error) {
+			} else {
+				return admission.NewForbidden(attr, fmt.Errorf("mutated spec.source.sourceSecret, expected: %v, got %v", api.LocalObjectReference{Name: secretName}, bc.Spec.Source.SourceSecret))
+			}
+		}
```
Isn't this changing behavior by suddenly forbidding stuff during Validation that was previously admitted ?

sttts replied:
It shouldn't be as long as nobody did dirty post-secretInjector mutation in admission webhooks, which we have disabled by default.

sttts commented:
/hold cancel

@liggitt's question is resolved.

deads2k commented on pkg/build/apiserver/admission/secretinjector/admission.go:
```diff
@@ -93,7 +104,13 @@ func (si *secretInjector) Admit(attr admission.Attributes) (err error) {
 		secretName := match.Cookie.(string)
 		glog.V(4).Infof(`secretinjector: matched secret "%s/%s" to buildconfig "%s"`, namespace, secretName, bc.GetName())
-		bc.Spec.Source.SourceSecret = &api.LocalObjectReference{Name: secretName}
+		if bc.Spec.Source.SourceSecret == nil || bc.Spec.Source.SourceSecret.Name != secretName {
```
Pretty sure this method returns early if `bc.Spec.Source.SourceSecret != nil`, so the first predicate is extraneous

deads2k replied:
and if it's always nil, then the second predicate will always npe.

I'm not seeing a job here for validation admission in this plugin.

sttts replied:
The func returns early on non-nil.

But on nil we still have to check whether `urlpattern.Match(patterns, url)` matches and then return the error during validation (a webhook could have reset the secret to nil).

The more fundamental question is whether we want to re-check at all in injection admission plugins. They are not security critical.

sttts commented on pkg/quota/apiserver/admission/runonceduration/admission.go:
```diff
@@ -90,17 +100,36 @@ func (a *runOnceDuration) Admit(attributes admission.Attributes) error {
 	}
 
-	appliedProjectLimit, err := a.applyProjectAnnotationLimit(attributes.GetNamespace(), pod)
+	newSpec := pod.Spec
```
done

sttts commented:
@deads2k addressed comments.

liggitt commented:
/unassign

sttts commented:
/retest

sttts commented:
> Can you comment on behavioral changes, what is expected, what is not expected ?

@simo5  we split the admission plugins into a mutating and a validating phase (we introduced that split in https://github.com/kubernetes/kubernetes/pull/54689 upstream, this PR does the corresponding work for all OpenShift admission plugins). This change allows us to call out to user defined admission webhook (compare https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks), let them mutate objects and then at the end our admission plugins' validation is called. So we can be sure that admission is effective even though users can customize it.

So in short: behaviour should be changed of any of the plugins. Instead the calls to `Admit`+webhook+`Validate` should do the same as am `Admit` before. If the webhook mutates what `Admit` did, the `Validation` call should catch that and refuse.

sttts commented:
Rebased.

sttts commented:
@simo5 lgty?

sttts commented:
/retest

simo5 commented:
/lgtm

sttts commented:
@deads2k with jordan removed from OWNERS this lost the approval.

deads2k commented:
/approve
/retest

sttts commented:
Looks like the gcp test failure is real: `StdErr: "Error from server (Forbidden): pods \"multicast-0\" is forbidden: exec operation is not allowed because the pod's security context exceeds your permissions: pods \"multicast-0\" is forbidden: unable to validate against any pod security policy: []"`

sttts commented:
@simo5 @deads2k please check the SCCExecRestrictions commit. `SCC.Validate` is very strict in the sense that the SA/user being checked for exec would have access to the SCC that was chosen when creating the pod. `SCC.Admit` in contrast allow a switch of the SCC (which is stored in the annotation).

sttts commented:
/retest

sttts commented:
@deads2k @simo5 lgty?

simo5 commented:
/lgtm

sttts commented:
/cherrypick release-3.11
